### PR TITLE
Improve wle reg/unreg robustness

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
- When `RegisterWorkload` fails(maybe workloadgroup not exist), close the stream and make retry originated from sidecar. This is to make istiod stateless in this case.

- use patch instead of update for unregister

- make connect/disconnect time readable

### CODE REVIEW

File: pilot/pkg/xds/workloadentry.go [hunk 1]
```diff
@@ -101,15 +107,11 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
-	if cfg == nil {
-		// we failed to create the workload entry in the first place
-		return
-	}
-	wle := cfg.DeepCopy()
-	delete(wle.Annotations, WorkloadControllerAnnotation)
-	wle.Annotations[DisconnectedAtAnnotation] = strconv.FormatInt(time.Now().UnixNano(), 10)
-	_, err := sg.Store.Update(wle)
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		delete(cfg.Annotations, WorkloadControllerAnnotation)
```
Comment (howardjohn) [inline pilot/pkg/xds/workloadentry.go:111]: I think patch here is intended so if another Istiod' has already taken ownership we do not overwrite.

Comment (stevenctl) [reply pilot/pkg/xds/workloadentry.go:111]: Right about the "so we don't overwrite", but we want to use "Update" since that fails on conflict.

**Comment by stevenctl** [comment]
lgtm except for using Patch in the unregister flow. That's probably what's failing the test.

**Comment by hzxuzhonghu** [comment]
The ut failure is because the time format precision is not enough.  Not related to patch

**Review by adiprerepa — APPROVED** [review]

**Comment by stevenctl** [comment]
The race still stands.. when disconnecting and reconnecting to a different pilot, we may end up un-setting the label.. we need to use update on unregister

**Comment by hzxuzhonghu** [comment]
>The race still stands.. when disconnecting and reconnecting to a different pilot, we may end up un-setting the label.

I do not see any bad result, the "istio.io/disconnectedAt" is just a indication the latest disconnect timestamp. When it reconnects to a new pilot,  this annotation is always kept, and it does not influence the cleanup either

**Comment by stevenctl** [comment]
The delayed cleanup logic checks if we have the controller annotation set. That patch will unset that label.

**Comment by hzxuzhonghu** [comment]
>The delayed cleanup logic checks if we have the controller annotation set. That patch will unset that label.

Hmm, you mean `"istio.io/workloadController"`.

**Comment by stevenctl** [comment]
Yes. Eventually we won't need to worry about that.. we've talked about doing a heartbeat based implementation that won't require it.  It handles an edge case where istiod crashes and the associated proxy never re-connects, we'll have a hanging WorkloadEntry. See a comment from Shriram in the [design doc](https://docs.google.com/document/d/10TyPy-6bOs0_7JljGpc2RneDdsaR9GktjwwaaVbz6fs/edit) if you're curious. 

As far as this PR is concerned, we do use the workloadController annotation to decide whether or not to cleanup and use it to indicate that the WorkloadEntry has its associated proxy still connected to some istiod instance.

**Comment by hzxuzhonghu** [comment]
Yeah, Let me fix it. Should be easy, we can check if this wle belongs to me, i will do cleanup. Otherwise, leaves it.

**Review by stevenctl — APPROVED** [review]

**Pull request merged**